calico_kubernetes/tests/policy_test.py

```diff
@@ -0,0 +1,183 @@
+
+        # Parse the rule
+        parsed = self.parser.parse_line(rule)
+
+        # assert on the result
+        self.assertEqual(parsed['action'], "allow")
+        self.assertEqual(parsed['protocol'], "udp")
+        self.assertEqual(parsed['src_ports'], [443])
+        self.assertEqual(parsed['src_net'], IPNetwork("1.2.3.4/32"))
+        self.assertEqual(parsed['dst_ports'], [80, 90, 100])
+
+        # Again
+        rule = "allow udp to cidr 1.2.0.0/16"
+
+        # Parse the rule
+        parsed = self.parser.parse_line(rule)
+
+        # Assert on the result
+        self.assertEqual(parsed['action'], "allow")
+        self.assertEqual(parsed['protocol'], "udp")
+        self.assertEqual(parsed['dst_net'], IPNetwork("1.2.0.0/16"))
+
+    def test_parse_tcp_rule(self):
+        rule = "allow tcp from label test=label to ports 80,90,100"
+
+        # Parse the rule
+        parsed = self.parser.parse_line(rule)
+
+        # Assert on the result
+        expected_tag = "%s_test_label" % self.namespace
+        self.assertEqual(parsed['action'], "allow")
+        self.assertEqual(parsed['protocol'], "tcp")
+        self.assertEqual(parsed['src_tag'], expected_tag)
+        self.assertEqual(parsed['dst_ports'], [80, 90, 100])
+
+        # Try again with a destination label.
+        rule = "allow tcp to ports 80,90,100 label test=label"
+
+        # Parse the rule
+        parsed = self.parser.parse_line(rule)
+
+        # Assert on the result
+        expected_tag = "%s_test_label" % self.namespace
+        self.assertEqual(parsed['action'], "allow")
+        self.assertEqual(parsed['protocol'], "tcp")
+        self.assertEqual(parsed['dst_tag'], expected_tag)
+        self.assertEqual(parsed['dst_ports'], [80, 90, 100])
+
+    def test_parse_icmp_rule(self):
+        rule = "allow icmp type 8 code 2 from label A=B to cidr 1.2.0.0/16"
+
+        # Parse the rule
+        parsed = self.parser.parse_line(rule)
+
+        # Assert on the result
+        expected_tag = "%s_A_B" % self.namespace
+        self.assertEqual(parsed['action'], "allow")
+        self.assertEqual(parsed['protocol'], "icmp")
+        self.assertEqual(parsed['icmp_type'], 8)
+        self.assertEqual(parsed['icmp_code'], 2)
+        self.assertEqual(parsed['src_tag'], expected_tag)
+        self.assertEqual(parsed['dst_net'], IPNetwork("1.2.0.0/16"))
+
+        rule = "allow icmp type 8 from cidr 1.2.0.0/16"
+
+        # Parse the rule
+        parsed = self.parser.parse_line(rule)
+
+        # Assert on the result
+        self.assertEqual(parsed['action'], "allow")
+        self.assertEqual(parsed['protocol'], "icmp")
+        self.assertEqual(parsed['icmp_type'], 8)
+        self.assertEqual(parsed['src_net'], IPNetwork("1.2.0.0/16"))
+
+    def test_parse_errors(self):
+        # Missing icmp type.
+        rule = "allow icmp type code 4"
+        self.assertRaises(PolicyParseError, self.parser.parse_line, rule)
```

Comment:
This code would be more concise IMO if you used the @parameterized decorator. You'd trim n spaces, and n assertRaises.